packages/bitcore-wallet-service/src/lib/server.ts

```diff
@@ -9286,214 +9303,313 @@ export class WalletService {
-    //     "barcode": "00000002104",
-    //     "claimed": "false"
-    //   },
-    // ]
-
+    // Filter to get list Device active (countNumber > 0)
     const opts = {
       isActive: true
     };
 
-    async.series(
+    let newListDeviceLow, newListDeviceMedium, newListDeviceHigh;
+
+    async.waterfall(
       [
         next => {
-          this.storage.fetchAllLogDevice(opts, (err, d) => {
+          this.storage.fetchAllLogDevice(opts, (err, listDevice) => {
             if (err) {
               return next(err);
             }
-            listDevice = d;
-            next();
+            if (listDevice) return next(null, listDevice);
           });
         },
-        next => {
-          if (listDevice) {
-            listDevice.forEach(device => {
-              if (device.countNumber > 0 && device.countNumber <= 3) {
-                listDeviceLow.push(device);
-              } else if (device.countNumber > 3 && device.countNumber <= 6) {
-                listDeviceMedium.push(device);
-              } else if (device.countNumber === 7) {
-                listDeviceHigh.push(device);
-              }
-            });
-            next();
-          }
-        },
-        next => {
-          if (listDeviceLow.length === listCodeWeeklyLowCsv.length) {
-            listDeviceLow.map((device, i) => {
-              listCodeWeeklyLowCsv.map((voucher, j) => {
-                if (i == j && voucher.claimed === 'false') {
-                  let appreciation = Appreciation.create({
-                    deviceId: device.deviceId,
-                    claimCode: voucher.claimCode,
-                    type: 'Weekly'
+        (listDevice, next) => {
+          let listDeviceLow = [], listDeviceMedium = [], listDeviceHigh = [];
+          let listCodeWeeklyLowCsv = [], listCodeWeeklyMediumCsv = [], listCodeWeeklyHighCsv = [];
+          // After filter, have list Device => group list device by countNumber.
+          listDevice.forEach(device => {
+            if (device.countNumber > 0 && device.countNumber <= 3) {
+              listDeviceLow.push(device);
+            } else if (device.countNumber > 3 && device.countNumber <= 6) {
+              listDeviceMedium.push(device);
+            } else if (device.countNumber === 7) {
+              listDeviceHigh.push(device);
+            }
+          });
+          // Read file csv => Have a group claim code data.
+          this.readDataCvsWeekly((err, listClaimCodeCsv) => {
+            if (err) return next(err);
+            if (listClaimCodeCsv) {
+              listCodeWeeklyLowCsv = listClaimCodeCsv[0];
+              listCodeWeeklyMediumCsv = listClaimCodeCsv[1];
+              listCodeWeeklyHighCsv = listClaimCodeCsv[2];
+              
+              if (listDeviceLow.length === listCodeWeeklyLowCsv.length) {
+                newListDeviceLow = _.clone(listDeviceLow);
+                newListDeviceLow.map((device, i) => {
+                  listCodeWeeklyLowCsv.map((lixi, j) => {
```

Comment:
For this compare we can use _.zip()
It will match items in 2 arrays with same index.